hasher-matcher-actioner/hmalib/lambdas/hashing.py

```diff
@@ -71,30 +78,46 @@ def lambda_handler(event, context):
         if message.get("Event") == "s3:TestEvent":
             continue
 
+        media_to_process: t.List[t.Union[S3ImageSubmission, URLSubmissionMessage]] = []
+
         if URLSubmissionMessage.could_be(message):
-            submission_message = URLSubmissionMessage.from_sqs_message(message)
+            media_to_process.append(URLSubmissionMessage.from_sqs_message(message))
+        elif S3ImageSubmissionBatchMessage.could_be(message):
+            # S3 submissions can only be images for now.
+            media_to_process.extend(
+                S3ImageSubmissionBatchMessage.from_sqs_message(
+                    message, image_prefix=IMAGE_PREFIX
+                ).image_submissions
+            )
         else:
             logger.warn(f"Unprocessable Message: {message}")
 
-        content_type = get_content_type_for_name(submission_message.content_type)
-        if not hasher.supports(content_type):
-            logger.warn(f"Unprocessable content type: {content_type}")
-            continue
-
-        with metrics.timer(metrics.names.hasher.download_file):
-            bytes_: bytes = URLContentSource().get_bytes(submission_message.url)
-
-        for signal in hasher.get_hashes(
-            submission_message.content_id, content_type, bytes_
-        ):
-            hash_record = PipelineHashRecord(
-                content_id=submission_message.content_id,
-                signal_type=signal.signal_type,
-                content_hash=signal.signal_value,
-                updated_at=datetime.datetime.now(),
-            )
-
-            hasher.write_hash_record(records_table, hash_record)
-            hasher.publish_hash_message(sqs_client, hash_record)
+        for media in media_to_process:
+            if not hasher.supports(media.content_type):
+                logger.warn(f"Unprocessable content type: {media.content_type}")
+                continue
+
+            with metrics.timer(metrics.names.hasher.download_file):
+                if hasattr(media, "key") and hasattr(media, "bucket"):
```

Comment:
confused: you already did/had disambiguation above? this check feels like more work due to losing some context.

Reply:
Nope, the disambiguation was between `URLSubmissionMessage` and `S3ImageSubmissionBatchMessage`. For S3ImageSubmissionBatchMessage, we end up adding a property `.image_submissions` to the `media_to_process` list. 

Here, we are differentiating between `URLSubmissionMessage` and `S3ImageSubmission` which is a different class.

Reply:
So I could be missing something but if I know something is an `URLSubmissionMessage` or a `S3ImageSubmission` (from `S3ImageSubmissionBatchMessage`) I then know how to get its bytes right there and then but instead we add them to the same list and do another check?

The two different message object are already disambiguated by their type, right?

Meta question that will explain my point or make it clear I am misunderstanding: 
Could media_to_process be created as a  list of tuples[message, get_bytes_method] ?

@schatten Thanks!

Reply:
That does make sense. We could reduce the checks!

Action item: @schatten don't let this slip your mind. :)